feat(plugins): add LSP integration plugin - Phase 1 diagnostics

TimeToBuildBob opened the pull request:
## Summary

Implements Issue gptme/gptme#961 - LSP support for gptme.

## Phase 1 Features

This initial release focuses on **diagnostics** - the most reliable and actionable LSP feature for AI assistants.

### Tool Commands
- `lsp diagnostics <file>` - Get errors/warnings for Python/TypeScript files
- `lsp status` - Show available language servers
- `lsp check` - Run diagnostics on all changed files (git)

### Post-Save Hook
Automatically shows errors after saving files - catch issues immediately after edits.

### Supported Languages
| Language | Tool Required | Install Command |
|----------|--------------|-----------------|
| Python | pyright | `npm install -g pyright` |
| TypeScript/JavaScript | tsc | `npm install -g typescript` |

## Architecture

```txt
gptme_lsp/
├── __init__.py          # Plugin entry point
├── lsp_client.py        # LSP protocol client (for future phases)
├── tools/
│   ├── __init__.py
│   └── lsp_tool.py      # LSP tool implementation
└── hooks/
    ├── __init__.py
    └── post_save.py     # Post-save diagnostics hook
```

## Future Phases

- **Phase 2**: Go-to-definition, find references, hover documentation
- **Phase 3**: Auto-fix suggestions, refactoring support

## References

- Issue: gptme/gptme#961
- Design inspiration: [OpenCode](https://opencode.ai/) - terminal-based AI coding agent with LSP integration
- Related: [mcp-language-server](https://github.com/isaacphi/mcp-language-server) - MCP-based alternative

## Testing

```bash
# Install the plugin
pip install -e plugins/lsp

# Test diagnostics
lsp status
lsp diagnostics src/myfile.py
```
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Adds LSP integration plugin for gptme with diagnostics, server status, and post-save hooks, supporting Python and TypeScript/JavaScript.
> 
>   - **Behavior**:
>     - Adds `lsp diagnostics <file>`, `lsp status`, and `lsp check` commands in `lsp_tool.py` for diagnostics, server status, and checking changed files.
>     - Implements post-save diagnostics hook in `post_save.py` to automatically run diagnostics after saving files.
>   - **Architecture**:
>     - Introduces `LSPServer` and `LSPManager` classes in `lsp_client.py` to manage language server processes and diagnostics.
>     - Modular structure with `tools/lsp_tool.py` for tool implementation and `hooks/post_save.py` for hooks.
>   - **Testing**:
>     - Adds tests in `test_lsp_tool.py` for tool execution, server availability, and diagnostics functionality.
>     - Includes tests for command registration and execution.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=gptme%2Fgptme-contrib&utm_source=github&utm_medium=referral)<sup> for 3b13444aeb2d41e436a6f2d778051b102ed0ae7d. You can [customize](https://app.ellipsis.dev/gptme/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

TimeToBuildBob commented:
## ✅ Additional Fix (bc6dd81)

Addressed remaining Greptile review comment:

- **lsp_client.py:335**: Replaced `subprocess.run(["which", command])` with `shutil.which(command)` for cross-platform Windows compatibility

All cross-platform compatibility issues now resolved. The plugin should work correctly on Windows, macOS, and Linux.

TimeToBuildBob commented:
## LSP Protocol Implementation Progress ✅

@ErikBjare - Addressed your review feedback about proper LSP support:

### Changes Made (commit 0d40c48)

**1. lsp_client.py - Fixed Diagnostics Collection**
- Added `_diagnostics` storage field to `LSPServer`
- Updated `_handle_notification()` to properly parse and store diagnostics from `textDocument/publishDiagnostics` notifications
- Fixed `get_diagnostics()` to return stored diagnostics with proper polling/timeout

**2. lsp_tool.py - Generic LSP Support**
- Added `_get_or_start_server()` for generic LSP server management
- Added `_get_lsp_diagnostics()` that uses proper LSP protocol (not CLI)
- Updated `execute()` to try LSP first, with CLI as fallback
- Now supports any LSP-compliant server: pyright, typescript-language-server, gopls, rust-analyzer

### Review Comments Addressed

| Comment | Status |
|---------|--------|
| Line 74: 'hardcodes pyright/typescript specific code' | ✅ Fixed - uses generic LSP client |
| Line 159: 'not actual LSP support' | ✅ Fixed - uses proper LSP protocol |
| Line 200: Use proper hook type | 🔄 Planned for next commit |

### Still TODO

1. **Hook Type Fix**: Change from `TOOL_POST_EXECUTE` + tool name check to `FILE_POST_SAVE` hook type
2. **Testing**: Verify LSP servers work correctly in practice

The plugin now uses the Language Server Protocol generically rather than hardcoding specific language tool CLIs.

ErikBjare reviewed (commented):
Looks pretty good now. Will test it later.

Could you add a `/lsp` command as well? (either as ToolSpec param or as `commands/` module.

TimeToBuildBob commented:
Added the `/lsp` command as requested! ✅

The command is registered via ToolSpec's `commands` parameter. Usage:
- `/lsp` - Show status (default)
- `/lsp diagnostics <file>` - Get errors/warnings for a file
- `/lsp status` - Show available language servers
- `/lsp check` - Run diagnostics on all changed files

Also fixed a test that was importing the removed `_ensure_pyright` function.

Commit: 3b13444
